modules/core/src/test/java/org/apache/ignite/internal/processors/datastructures/IgniteReentrantLockTest.java

```diff
@@ -0,0 +1,426 @@
+
+            lock.unlock();
+
+            assertFalse(lock.isLocked());
+            assertFalse(lock.isHeldByCurrentThread());
+        }
+    }
+
+    /**
+     * Test an async lock acquiring and releasing with many nodes.
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinode() throws Exception {
+        testReentrantLockMultinode(false);
+        testReentrantLockMultinode(true);
+    }
+
+    /**
+     * {@link this#testReentrantLockMultinode()}
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinode(final boolean fair) throws Exception {
+        final AtomicInteger inx = new AtomicInteger(0);
+
+        GridTestUtils.runMultiThreadedAsync(new Runnable() {
+            @Override public void run() {
+                try {
+                    IgniteLock lock = createReentrantLock(inx.getAndIncrement(), fair ? "lock1" : "lock2", fair);
+                    lock.lock();
+
+                    try {
+                        assertTrue(lock.isLocked());
+                        Thread.sleep(1_000L);
+                    }
+                    catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                    finally {
+                        lock.unlock();
+                    }
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
+                }
+            }
+        }, NODES_CNT, "worker").get(30_000L);
+    }
+
+    /**
+     * {@link this#testReentrantLockMultinodeTimeout(boolean)}
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeTimeout() throws Exception {
+        testReentrantLockMultinodeTimeout(false);
+        testReentrantLockMultinodeTimeout(true);
+    }
+
+    /**
+     * {@link this#testReentrantLockMultinode()} with timeout.
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeTimeout(final boolean fair) throws Exception {
+        final AtomicInteger inx = new AtomicInteger(0);
+
+        GridTestUtils.runMultiThreadedAsync(new Runnable() {
+            @Override public void run() {
+                try {
+                    IgniteLock lock = createReentrantLock(inx.getAndIncrement(), fair ? "lock1" : "lock2", fair);
+
+                    if (lock.tryLock(10, TimeUnit.NANOSECONDS)) {
+                        try {
+                            assertTrue(lock.isLocked());
+                            Thread.sleep(1_000L);
+                        }
+                        catch (InterruptedException e) {
+                            e.printStackTrace();
+                        }
+                        finally {
+                            lock.unlock();
+                        }
+                    }
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
```

Comment:
mb, its better to let the exception outside

Reply:
Runnable.run() doesn't throw any exceptions.

Reply:
so may be use callable instead

Reply:
fixed